packages/react/src/components/NumberInput/NumberInput.tsx

```diff
@@ -1073,9 +1169,26 @@ const HelperText = ({ disabled, description, id }: HelperTextProps) => {
  * @param {number} config.value
  * @param {number} config.max
  * @param {number} config.min
+ * @param {Function} config.validate
+ * @param {string} config.locale
  * @returns {boolean}
  */
-function getInputValidity({ allowEmpty, invalid, value, max, min }) {
+function getInputValidity({
+  allowEmpty,
+  invalid,
+  value,
+  max,
+  min,
+  validate,
+  locale,
+}) {
+  if (typeof validate === 'function') {
+    const result = validate(value, locale);
+    if (result !== undefined) {
```

Comment:
I would say that if we are invalid we can return immediately.  But if we are "valid" here, we could still become invalid later.  ie, min/max, etc.  I see the validator does specifiy that there are 3 returns true/false/undefined but the number validation functions return 'true' for things like 'empty number', so what if things are required?

To keep things simple, I would focus on a false return as being invalid and return immediately.  Any `true` should just allow for further validation.

Unless the intention really is that the validation function should supercede any validation such as min/max, require, etc.  And if that is the case then we should update the locale parsing validation functions to return undefined where they currently return true.